libs/s25main/EconomyModeHandler.h

```diff
@@ -0,0 +1,93 @@
+#pragma once
+#include "GameObject.h"
+#include "gameTypes/GoodTypes.h"
+#include "gameTypes/Inventory.h"
+#include "gameTypes/JobTypes.h"
+#include "gameData/MaxPlayers.h"
+
+#include <vector>
+
+class SerializedGameData;
+class GameEvent;
+
+// Handler object to keep track of the economy mode progress and for the game end event
+class EconomyModeHandler : public GameObject
+{
+public:
+    static const unsigned int numGoodTypesToCollect = 7;
+
+    struct econTeam
+    {
+        unsigned mask;
+        unsigned teamAmounts[numGoodTypesToCollect];
+        unsigned num_players_in_team;
+        unsigned teamWins;
+
+        econTeam(unsigned mask, unsigned num_players_in_team) noexcept
+            : mask(mask), num_players_in_team(num_players_in_team)
+        {
+            for(unsigned int& teamAmount : teamAmounts)
+            {
+                teamAmount = 0;
+            }
+            teamWins = 0;
+        }
+
+        bool inTeam(unsigned int playerId) const;
+    };
+
+private:
+    /// Frame in which the game is going to end
+    unsigned end_frame;
+    /// End game Event
+    const GameEvent* event;
+    /// Good types to collect
+    GoodType types[numGoodTypesToCollect];
+
+    // Data for economy mode progress tracking
+    std::vector<econTeam> teams;
+    unsigned maxTeamAmounts[numGoodTypesToCollect] = {0};
+    unsigned int mostWins = 0;
+    unsigned amounts[numGoodTypesToCollect][MAX_PLAYERS];
+    unsigned last_updated;
```

Comment:
The amounts are only updated when either iwEconomicProgress or the end game event request it. I save the gameframe at which we updated last, because iwEconomicProgress asks for an update every video frame and otherwise we would update more than once in a gameframe.

Reply:
Makes sense, alright :)